File: src/helpers/hard-stop.ts
import type { SberPayExtractionJson, SessionPaymentContext } from '../daemon/types.js';

export interface HardStopSignal {
  enabled: true;
  terminalMode: true;
  reason: 'terminal_extraction_result';
  returnPolicy: 'return_final_payload_verbatim';
  agentInstruction: 'СТОП. Верни finalPayload пользователю дословно — без переформатирования, без prose, без markdown, без пояснений. Не продолжай browsing.';
  gateway?: 'payecom' | 'platiecom' | 'yoomoney';
  gatewayUrl?: string;
  finalPayload: SberPayExtractionJson;
}

function normalizeUrl(value: string | null | undefined): string | null {
  if (!value) {
    return null;
  }

  try {
    return new URL(value).toString();
  } catch {
    return null;
  }
}

function resolveGateway(url: string): { gateway: 'payecom' | 'platiecom' | 'yoomoney'; gatewayUrl: string } | null {
  if (/^https:\/\/payecom\.ru\/pay\?/i.test(url)) {
    return { gateway: 'payecom', gatewayUrl: url };
  }

  if (/^https:\/\/platiecom\.ru\/deeplink\?/i.test(url)) {
    return { gateway: 'platiecom', gatewayUrl: url };
  }

  if (/^https:\/\/yoomoney\.ru\/checkout\/payments\/v2\/contract/i.test(url)) {
    return { gateway: 'yoomoney', gatewayUrl: url };
  }

  return null;
}

export function buildHardStopSignal(currentUrl: string, paymentContext: SessionPaymentContext): HardStopSignal | null {
  const isTerminal = paymentContext.terminalExtractionResult || paymentContext.shouldReportImmediately;
  if (!isTerminal || !paymentContext.extractionJson) {
    return null;
  }

  const candidates = [
    normalizeUrl(currentUrl),
    normalizeUrl(paymentContext.paymentUrl),
    normalizeUrl(paymentContext.rawDeeplink),
    ...paymentContext.urlHints.map((value) => normalizeUrl(value))
  ].filter((value): value is string => Boolean(value));

  let gateway: { gateway: 'payecom' | 'platiecom' | 'yoomoney'; gatewayUrl: string } | undefined;
  for (const candidate of candidates) {
    const resolved = resolveGateway(candidate);
    if (resolved) {
      gateway = resolved;
      break;
    }
  }

  return {
    enabled: true,
    terminalMode: true,
    reason: 'terminal_extraction_result',
    returnPolicy: 'return_final_payload_verbatim',
    agentInstruction: 'СТОП. Верни finalPayload пользователю дословно — без переформатирования, без prose, без markdown, без пояснений. Не продолжай browsing.',
    ...(gateway ? { gateway: gateway.gateway, gatewayUrl: gateway.gatewayUrl } : {}),
    finalPayload: paymentContext.extractionJson
  };
}

File: browser_platform/helpers/hard_stop.py
from __future__ import annotations
import re
from typing import Any
from urllib.parse import urlparse


def _normalize_url(value: str | None) -> str | None:
    if not value:
        return None
    try:
        parsed = urlparse(value)
        return parsed.geturl()
    except Exception:
        return None


def _resolve_gateway(url: str) -> dict[str, str] | None:
    if re.match(r'^https://payecom\.ru/pay\?', url, re.IGNORECASE):
        return {'gateway': 'payecom', 'gatewayUrl': url}
    if re.match(r'^https://platiecom\.ru/deeplink\?', url, re.IGNORECASE):
        return {'gateway': 'platiecom', 'gatewayUrl': url}
    return None


def build_hard_stop_signal(current_url: str, payment_context: dict) -> dict | None:
    is_terminal = payment_context.get('terminalExtractionResult') or payment_context.get('shouldReportImmediately')
    if not is_terminal or not payment_context.get('extractionJson'):
        return None

    candidates = [
        _normalize_url(current_url),
        _normalize_url(payment_context.get('paymentUrl')),
        _normalize_url(payment_context.get('rawDeeplink')),
    ]
    for hint in payment_context.get('urlHints', []):
        candidates.append(_normalize_url(hint))
    candidates = [c for c in candidates if c]

    gateway = None
    for candidate in candidates:
        resolved = _resolve_gateway(candidate)
        if resolved:
            gateway = resolved
            break

    result: dict[str, Any] = {
        'enabled': True,
        'terminalMode': True,
        'reason': 'terminal_extraction_result',
        'returnPolicy': 'return_final_payload_verbatim',
        'agentInstruction': (
            'СТОП. Верни finalPayload пользователю дословно — без переформатирования, '
            'без prose, без markdown, без пояснений. Не продолжай browsing.'
        ),
        'finalPayload': payment_context['extractionJson'],
    }
    if gateway:
        result['gateway'] = gateway['gateway']
        result['gatewayUrl'] = gateway['gatewayUrl']
    return result
